### ISSUE TITLE
SC2004 ((${!a})) dollar in variable indirection is necessary in expressions.

### ISSUE THREAD

**Issue by informatimago** [koalaman/shellcheck#891]
#### For new checks and feature suggestions
- [v] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [v] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
b=42
a=b
((${!a}++))
```

#### Here's what shellcheck currently says:
On the `((${!a}++))` line:
```
$/${} is unnecessary on arithmetic variables. [SC2004]
```

#### Here's what I wanted or expected to see:
Nothing. In (( ... )), ${!a} is not redundant.

**Comment by kurahaupo** [comment]
You should note that variables are recursively expanded as subexpressions as R-values, but are simply overwritten upon assignment, including by the `++` operator.

It might not be redundant _in general_, but in that code it's certainly wrong:

    ( b=42 a=b ; ((${!a}++)) )
    -bash: ((: 42++: syntax error: operand expected (error token is "+")

Which is obvious when you consider that you don't normally write `$x=1`, so `${!x}=1` is all the more unusual. What's happening is that you're attempting _two_ levels of indirection off `a`, which is unlikely to be what you want:

    ( a=5 b=a c=b ; ((c*=2)) ; echo $a $b $c )
    5 a 10

    ( a=5 b=a c=b ; (($c*=2)) ; echo $a $b $c )
    5 10 b

    ( a=5 b=a c=b ; ((${!c}*=2)) ; echo $a $b $c )
    10 a b

**Comment by koalaman** [comment]
This was fixed in 8e5e77ad7. Thanks!
